### ISSUE TITLE
insert result to sequence column is incorrect

### ISSUE THREAD

**Issue by yanwr1** [greenplum-db/gpdb#15579]
## Bug Report

After rollback a transaction, the insert result to sequence column is incorrect.
### Greenplum version or build
PostgreSQL 12.12 (Greenplum Database 7.0.0-alpha.0+dev.18551.g9a365ec8432 build dev)
### OS version and uname -a
x86_64-pc-linux-gnu
### Step to reproduce the behavior
```
create sequence t_a cache 20;
create table t(a int default nextval('t_a'));
ALTER SEQUENCE t_a OWNED BY t.a;

INSERT INTO t DEFAULT VALUES;
INSERT INTO t DEFAULT VALUES;
postgres=# select * from t;
 a 
---
 1
 2
(2 rows)

begin;
TRUNCATE t RESTART IDENTITY;
INSERT INTO t DEFAULT VALUES;
postgres=# select * from t;
 a 
---
 1
(1 row)
rollback;

INSERT INTO t DEFAULT VALUES;
postgres=# select * from tt;
 a 
---
 2
 2
 1
(3 rows)
```

**Comment by yanwr1** [comment]
insert result in postgres:
```
postgres=# select * from t;
 a  
----
  1
  2
 21
(3 rows)
```

**Comment by yanwr1** [comment]
the insert plan:
```
postgres=# explain (analyze, verbose) INSERT INTO t DEFAULT VALUES;
                                                        QUERY PLAN                                                         
---------------------------------------------------------------------------------------------------------------------------
 Insert on public.t  (cost=0.00..0.01 rows=1 width=4) (actual time=0.000..0.804 rows=0 loops=1)
   ->  Result  (cost=0.00..0.00 rows=1 width=8) (actual time=0.623..0.624 rows=1 loops=1)
         Output: (int4(nextval('t_a'::regclass)))
         ->  Redistribute Motion 1:3  (slice1)  (cost=0.00..0.00 rows=1 width=4) (actual time=0.621..0.621 rows=1 loops=1)
               Output: (int4(nextval('t_a'::regclass)))
               Hash Key: (int4(nextval('t_a'::regclass)))
               ->  Result  (cost=0.00..0.00 rows=1 width=1) (actual time=0.315..0.316 rows=1 loops=1)
                     Output: int4(nextval('t_a'::regclass))
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 29.795 ms
   (slice0)    Executor memory: 37K bytes avg x 3 workers, 37K bytes max (seg0).
   (slice1)    Executor memory: 41K bytes (entry db).
 Memory used:  128000kB
 Execution Time: 12.899 ms
(14 rows)
```
During the aborted transaction, the insert statement create a reader gang to acquire nextval value.And the qe process in the reader gang creates a hash table for storing sequence data, and insert the new hash table entry for sequence `t_a` to the hash table.
```
        /* Find or create a hash table entry for this sequence */
	if (seqhashtab == NULL)
		create_seq_hashtable();

	elm = (SeqTable) hash_search(seqhashtab, &relid, HASH_ENTER, &found);
```
gdb debug, print the current value of hash table entry
```
(gdb) p *elm
$2 = {key = {relid = 32796, called_from_dispatcher = false}, filenode = 0, lxid = 284, last_valid = true, last = 1, 
  cached = 20, increment = 1}
```
After rollback the transaction,  the previous reader gang wasn't cleaned, and then execute the insert statement, the hashtable is not cleaned, the same qe process will acquire the nextval value from cached numbers directly, though the hashtable entry is out of date.

---
the sequence hash table info need to be cleaned after abort the transaction.

**Comment by hyongtao-db** [comment]
Good analysis. Thanks a lot.
Seems strong related with [issue 15142](https://github.com/greenplum-db/gpdb/issues/15142) and [issue 14851](https://github.com/greenplum-db/gpdb/issues/14851)

**Comment by adam8157** [comment]
#15204 left the case `TRUNCATE ... RESTART IDENTITY` in `regress/truncate` unchanged, which is related to this issue and needs to be fixed with this issue.

**Comment by yanwr1** [comment]
> The main branch is fixed by #16031 Will 6X have the issue? @yanwr1

6X also have the issue, I will create a pr in 6x branch.

**Comment by kainwen** [comment]
Close this. Decide not to backport until there is a real customer report.
